mtp_api/apps/mtp_auth/validators.py

```diff
@@ -1,16 +1,31 @@
+from django.contrib.auth import get_user_model
 from oauth2_provider.oauth2_validators import OAuth2Validator
 
-from .models import ApplicationUserMapping
+from .models import ApplicationUserMapping, FailedLoginAttempt
 
 
-class ApplicationRequestValidator(OAuth2Validator):
+class LockedOut(Exception):
+    pass
+
 
+class ApplicationRequestValidator(OAuth2Validator):
     def validate_user(self, username, password, client, request, *args, **kwargs):
-        valid = super(ApplicationRequestValidator, self).validate_user(
-            username, password, client, request, *args, **kwargs
-        )
-        if valid and ApplicationUserMapping.objects.filter(
-                user=request.user, application=client).exists():
-            return True
+        user = get_user_model().objects.filter(username=username).first()
+        try:
+            if user and FailedLoginAttempt.is_locked_out(user, client):
+                raise LockedOut
+
+            valid = super().validate_user(
+                username, password, client, request, *args, **kwargs
+            )
+            if valid and ApplicationUserMapping.objects.filter(
+                    user=request.user, application=client).exists():
+                FailedLoginAttempt.delete_failed_attempts(user, client)
+                return True
+            elif user:
```

Comment:
What are your thoughts on locking by IP address as well as by username? The typical concern is that someone might choose X common passwords and then just bruteforce on the username instead of the password if one is just trying to get a valid log in. We probably won't have enough users for this to be a serious issue, though.

However, in the current implementation, as it only blocks on valid users, an attacker could brute force valid usernames by checking to see if they get locked out after 5 attempts.